experimental/Schemes/Auxiliary.jl

```diff
@@ -0,0 +1,175 @@
+@Markdown.doc """
+    pullback(f::CoveredSchemeMorphism)
+
+Return a function `phi` which takes any reasonable 
+argument `v` associated to the `codomain` of `f` and 
+produces ``f*(v)``.
+"""
+function pullback(f::CoveredSchemeMorphism)
+  X = domain(f)
+  Y = codomain(f)
+
+  function pullback_func(II::IdealSheaf)
+    scheme(II) === Y || error("ideal sheaf is not defined on the codomain of the function")
+    phi = covering_morphism(f)
+    ID = IdDict{AbsSpec, Ideal}()
+    for U in patches(domain(phi))
+      f_U = phi[U]
+      V = codomain(f_U)
+      pbf = pullback(f_U)
+      ID[U] = ideal(OO(U), pbf.(gens(II(V))))
+    end
+    return IdealSheaf(X, ID, check=false) # TODO: Set to false eventually.
+  end
+
+  function pullback_func(C::CartierDivisor)
+    phi = covering_morphism(f)
+    triv_dict = IdDict{AbsSpec, RingElem}()
+    OOX = OO(X)
+    # We need to do the following:
+    # The divisor D := f^* C needs to be trivialized on some refinement. 
+    # The one we're using consists of the preimages of the patches of 
+    # the `trivializing_covering` of `C`. These must be intersected 
+    # with each and every affine chart of X first. 
+    for U in patches(domain(phi))
+      V = codomain(phi[U])
+      for W in patches(trivializing_covering(C))
+        # In case V and W do not happen to lay in the same affine 
+        # chart, the preimage of W in U will be empty. So we can 
+        # skip the computation in that case.
+        success, par = _have_common_ancestor(V, W)
+        if success
+          # Reconstruct W as a PrincipalOpenSubset of the affine chart directly
+          # (Note that we might have gone through some calls to `simplify(...)`, 
+          # so this really is a non-trivial step)
+          inc_W_flat = _flatten_open_subscheme(W, par)
+          # To cheaply construct the preimage of W in U, just pull back the 
+          # complement equation.
+          h = complement_equation(codomain(inc_W_flat))
+          UW = PrincipalOpenSubset(U, pullback(phi[U])(OOX(par, V)(h)))
+          # Finally, we can add the trivialization on U ∩ f⁻¹(W) to our list.
+          triv_dict[UW] = pullback(phi[U])(first(C(V)))
+        end
+      end
+    end
+    return CartierDivisor(X, triv_dict) # We have to leave the checks back on, because local 
+                                        # equations might pull back to zero or zero divisors.
```

Comment:
Will the error message be helpful? I.e. that the pullback in this case is not defined? 
Otherwise the user will think it is a bug and blame us.

Reply:
I changed this.